implementations/rust/ockam/ockam_entity/src/imp.rs

```diff
@@ -0,0 +1,260 @@
+use crate::authentication::Authentication;
+use crate::history::ProfileChangeHistory;
+use crate::{
+    Contact, ContactsDb, EntityError, EventIdentifier, KeyAttributes, Profile, ProfileAuth,
+    ProfileChangeEvent, ProfileChanges, ProfileContacts, ProfileEventAttributes, ProfileIdentifier,
+    ProfileIdentity, ProfileSecrets, ProfileVault,
+};
+use ockam_vault_core::{PublicKey, Secret};
+
+#[derive(Clone)]
+pub struct ProfileImpl<V: ProfileVault> {
+    identifier: ProfileIdentifier,
+    pub(crate) change_history: ProfileChangeHistory,
+    contacts: ContactsDb,
+    pub(crate) vault: V,
+}
+
+impl<V: ProfileVault> ProfileImpl<V> {
+    pub fn new(
+        identifier: ProfileIdentifier,
+        change_events: Vec<ProfileChangeEvent>,
+        contacts: ContactsDb,
+        vault: V,
+    ) -> Self {
+        let profile = Self {
+            identifier,
+            change_history: ProfileChangeHistory::new(change_events),
+            contacts,
+            vault,
+        };
+
+        profile
+    }
+}
+
+impl<V: ProfileVault> ProfileImpl<V> {
+    /// Return clone of Vault
+    pub fn vault(&self) -> V {
+        self.vault.clone()
+    }
+}
+
+impl<V: ProfileVault> ProfileImpl<V> {
+    /// Generate fresh [`Profile`] update key and create new [`Profile`] using it
+    pub(crate) fn create_internal(
+        attributes: Option<ProfileEventAttributes>,
+        mut vault: V,
+    ) -> ockam_core::Result<Self> {
+        let prev_id = vault.sha256(Profile::NO_EVENT)?;
+        let prev_id = EventIdentifier::from_hash(prev_id);
+
+        let key_attributes = KeyAttributes::new(Profile::PROFILE_UPDATE.to_string());
+        let change_event = Self::create_key_event_static(
+            prev_id,
+            key_attributes.clone(),
+            attributes,
+            None,
+            &mut vault,
+        )?;
+
+        let change = ProfileChangeHistory::find_key_change_in_event(&change_event, &key_attributes)
+            .ok_or(EntityError::InvalidInternalState)?;
+        let public_key = ProfileChangeHistory::get_change_public_key(&change)?;
+
+        let public_kid = vault.compute_key_id_for_public_key(&public_key)?;
+        let public_kid = ProfileIdentifier::from_key_id(public_kid);
+
+        let profile = Self::new(public_kid, vec![change_event], Default::default(), vault);
```

Comment:
Let's just reassign over `public_key` for the name here instead of `public_kid`

Reply:
Why would we call public key id a public key?